src/coreclr/jit/compiler.cpp

```diff
@@ -5531,6 +5531,8 @@ int Compiler::compCompile(CORINFO_MODULE_HANDLE classPtr,
 #endif
 
     virtualStubParamInfo = new (this, CMK_Unknown) VirtualStubParamInfo(IsTargetAbi(CORINFO_NATIVEAOT_ABI));
+    m_preferredInitCctor = IsTargetAbi(CORINFO_NATIVEAOT_ABI) ? CORINFO_HELP_READYTORUN_NONGCSTATIC_BASE
```

Comment:
> Can we convert cctor trigger to NonGC static base when RyuJIT asks for this helper in getReadyToRunHelper? Would it affect anything?

I think this would be fine. We will be calling it with incorrect calling convention (assuming it returns `void`), but I don't see that causing any issues on any of our platforms.

> I assume RyuJIT already knows that cctor trigger helper generates the same side effect as the helpers to get one of the bases, so we would still be able to eliminate it (there's no codegen advantage in knowing these helpers would end up being same on NativeAOT).

The optimization is relying on CSE to eliminate the extraneous calls, so the exact return value has to match (and notably `CCTOR_TRIGGER` returns void and does not participate in CSE). But this is just the initialization, it is going to end up as one of the other helpers if we have field accesses that use those.

Reply:
>  CCTOR_TRIGGER returns void and does not participate in CSE

CCTOR_TRIGGER should participate in CSE. It is what happens today.

For example, run this with optimizations on and tiering disable:
```
TypeWithCtor.m(); // This adds cctor trigger
TypeWithCtor.m(); // This adds second cctor trigger

class TypeWithCtor
{
    static TypeWithCtor()
    {
         Console.WriteLine("TypeWithCtor");
    }

    public static void m()
    {
    }
}
```
JITed code in .NET 7 - notice that the redundant cctor trigger was eliminated: 
```
sub      rsp, 40
mov      rcx, 0x7FFBFACFBE10
mov      edx, 4
call     CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
nop
add      rsp, 40
ret
```

Reply:
> CCTOR_TRIGGER should participate in CSE. It is what happens today.

I mean the R2R `CCTOR_TRIGGER` helper. It is not going to participate in CSE after changing its node to be properly typed in this PR. I guess your example shows that this correctness fix can cause regressions in R2R. I don't see a simple way to resolve this unless we want to rely on how the emitter's GC reporting ends up working for this helper.

The normal static base helpers should indeed participate in CSE.

Reply:
Well, the simple way to resolve the above is to remove `CCTOR_TRIGGER` entirely, but I assume it needs some micro benchmarks first.

Reply:
For a small micro benchmark I see practically no difference between using `CCTOR_TRIGGER` or `STATIC_BASE_NON_GC`:
```csharp
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;

public class Program
{
    public static void Main()
    {
        C.M();

        Stopwatch timer = Stopwatch.StartNew();
        for (int i = 0; i < 1_000_000_000; i++)
        {
            C.M();
        }
        Console.WriteLine("{0}ms", timer.ElapsedMilliseconds);
        timer.Restart();
    }
}

public static class C
{
    static C()
    {
        Console.WriteLine("cctor");
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static void M()
    {
    }
}
```

crossgen'd with tiered compilation disabled I get the following on win-x64:
With CCTOR_TRIGGER: 2111ms to 2145ms in 10 runs
With STATIC_BASE_NON_GC: 2113ms to 2130ms in 10 runs

On linux-arm64 (rpi 400):
With CCTOR_TRIGGER: 17262ms
With STATIC_BASE_NON_GC: 17260ms 

(double checked with r2rdump that the right helpers were being called)

@BrianBohe, I think this means you should feel free to remove `CORINFO_HELP_READYTORUN_CCTOR_TRIGGER` as part of this PR and switch the JIT to just use `STATIC_BASE_NON_GC` for both R2R and NAOT.